### ISSUE TITLE
Battery drain much too high

### ISSUE THREAD

**Issue by lgd821** [ubports/ubuntu-touch#387]
- Device: Nexus 5
- Channel: stable
- Build: OTA-3

### Steps to reproduce
Disable WLAN, GPS and bluetooth, charge to 100% and wait 24 hours without using the phone. Remark: Is not an issue of having a bad or old accu.

### Expected behavoir
Charge level above 90%, equivalent to 240 hours standby time. Compared to Android 240 hours are not the best standby time possible with this phone

### Actual behavior
Charge level dropped down to <70% (>30% discharge within 24 hours)

**Comment by Flohack74** [comment]
Probably the kernel is not optimized. This will be a huge challenge, eventually kernel optimizations from CM oder other 3rd party ROMs could help.

**Comment by Guillerm2** [comment]
Device: Meizu Pro 5 Ubuntu Edition
Channel: stable
Build: OTA-3

I have the same problem since upgrade to OTA 3. 

Disable WLAN, GPS and Bluetooth, only 4G activated.
The last night, without using it (and without applications in the background), dropped the battery charge from 80% to 21% in six hours. 
This afternoon the charge dropped 40% in 4 hours, without using the display or apps (except 5 minutes).

**Comment by tihom7** [comment]
Device: Nexus 5
Channel: stable
Build: OTA-3

Same problem, the battery drains too quick. For 24 hours it is down completely without any usage of apps or even the display. Only WiFi and 4G.

**Comment by advocatux** [comment]
Device: Nexus 5
Channel: stable
Build: OTA-3

I've changed the battery in my N5 with a brand new original LG battery (with date August-2017) some weeks ago.
Disabling WLAN, GPS, and BT. Only 4G activated. Leaving the phone with 100% charge and no interactions with it goes to 5% charge in 9 hours.

**Comment by mihaelmilea** [comment]
device: turbo
channel: vivid devel, latest version
As of last(s) version(s) of the devel channel I started to experience as well a continuous drop line in the battery usage where before I used to have horizontal lines corresponding to the standby periods. I can see with terminal / top that the cpus are not switched off anymore, four of the cores are almost all the time online - before quite often I was seeing just one core online.
@Flohack74 - was anything changed in the kernel for the last versions of the vivid devel channel from the previous ones?

**Comment by Flohack74** [comment]
No that I know of maybe @rubencarneiro can take a look? We got complaints on all devices about reduced battery life btw ;)

**Comment by advocatux** [comment]
As I said before, since I bought the new battery for my N5, everyday it goes from 100% to 5% in just nine hours (leaving the phone alone). BUT today, and for the first time, it went from 100% to 66% in the same amount of time.

The only difference this time is that I left the phone to turn off by itself before recharging it. I already did it in this way in two others occasions but the result was the same as always. Third time's a charm?

It seems there's a well known problem [*] in how Android handles stats.bin percentage and how battery software management handles battery percentage, and you need to calibrate the battery. I believe there's a similar mismatch happening in UBports.

[*] In Spanish: http://www.mibqyyo.com/comunidad/discussion/comment/218912#Comment_218912

**Comment by simoxxx** [comment]
i found out that if you disable the screen rotation and reboot you will get a better battery life.. did you allready know?

**Comment by james-lyon** [comment]
Device: Aquaris E4.5
Channel: stable
Build: OTA-3

Similar reduced battery life after the recent update on my device

**Comment by derkleinemannvomfeldweg** [comment]
Is this probably the same issue which I reported for FP2, too? I reported it in #400 .

**Comment by NeoTheThird** [comment]
Since it's two different devices and the problem is not that present on other devices, let's track it in separate issues.

**Comment by simoxxx** [comment]
i think this problem is different from device to device.
And it depends a lot on how sensors and cpu are managed while the phone is in stand-by..

**Comment by mihaelmilea** [comment]
@NeoTheThird Yes, it seems to be a different issue for Pro 5 as disabling the rotation did not change anything. I opened a new issue: https://github.com/ubports/ubuntu-touch/issues/407

**Comment by delijati** [comment]
Same for BQ Aquaris E 4.5 if i deactivate mobile data its lasts 4 days as usual .. some service polling to much ??

**Comment by Flohack74** [comment]
Guys can you please open issue #415 and follow the config change that was made at the end to the push client config, and re-test everything? Thx!

**Comment by simoxxx** [comment]
we will have to put this on the balance too..
https://forums.ubports.com/topic/894/how-to-changing-default-cpu-governor/2

**Comment by Scym1** [comment]
The phone seems to have a longer battery life if I turn it off and on before leaving for the day. Some days, I don't restart it and I can notice faster battery drains.

**Comment by lgd821** [comment]
I've tested battery drain using different governors. All this tests are done with GPS, WLAN, Bluetooth and Data connection disabled. Phone not used or touched (= display off) during time of investigation. Every test run started with to 100% charged battery:

Governor "interactive": Battery drain = 56% in 22 hours => ~61% in 24 hours
Governor "conservative": Battery drain = 32% in 22 hours => ~35% in 24 hours (up/down thresholds unchanged)
Governor "powersave": Battery drain = 41% in 22 hours => ~45% in 24 hours

It seems that the selected governor has strong impact to battery drain, but all result are far away from battery drain when using Android. The standby time for the Nexus 5 having Android OS system is much higher.

Since Ubuntu is coming from PC/Notebook: Is it possible that hardware like e.g. WLAN seems to be disabled by software but WLAN hardware is still supplied consuming power from battery?
Is the GPU reduced in clocked or switched of whenever the display is off?

**Comment by Flohack74** [comment]
That looks really cool! The N5 one does not look so bad at all...

**Comment by mymike00** [comment]
IMHO the [link](https://forums.ubports.com/topic/894/how-to-changing-default-cpu-governor/2) that @simoxxx gave is interesting. what about a UI for changing the cpu governor?
or maybe a "battery saving" mode/option that could do some things like: switching to 4g to 3g, reduce a bit the brightness, change the cpu governor, ecc...

**Comment by paulmueller** [comment]
I would like to share my experience. Initially, I had about 2% of battery drain per hour with only the cellular radio on (no internet connection, bluetooth, wlan, or gps).

Using the various sources listed here and in the forum, I came up with an ok-ish workaround-solution:

```
phablet@fon5:~$ cat /home/phablet/scripts/governor.sh 
#!/bin/sh
# https://forums.ubports.com/topic/894/how-to-changing-default-cpu-governor
sudo su <<EOF
echo conservative > /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor
echo 40 > /sys/devices/system/cpu/cpufreq/conservative/down_threshold
echo 90 > /sys/devices/system/cpu/cpufreq/conservative/up_threshold
EOF
# kill bluetooth thing
# https://forums.ubports.com/topic/1958/n5-battery-usage-in-ut-and-in-android/20
# https://github.com/ubports/ubuntu-touch/issues/676
sudo pkill -9 hciattach
sudo pkill -9 hciattach
sudo pkill -9 hciattach  # not sure why I did this three times
exit 0
```

To apply the changes on boot, I made this script executable (``chmod 775 /home/phablet/scripts/governor.sh``) and added it to ``/etc/init.d/rc.local`` (You have to set the system image writable, e.g. using UT Tweak Tool):

```
phablet@fon5:~$ cat /etc/init.d/rc.local 
#! /bin/sh
### BEGIN INIT INFO
# Provides:          rc.local
# Required-Start:    $all
# Required-Stop:
# Default-Start:     2 3 4 5
# Default-Stop:
# Short-Description: Run /etc/rc.local if it exist
### END INIT INFO


PATH=/sbin:/usr/sbin:/bin:/usr/bin

. /lib/init/vars.sh
. /lib/lsb/init-functions

### I only added the following two lines here (the script must be executed before any exit-statement)
# set CPU governor to conservative
/home/phablet/scripts/governor.sh

do_start() {
	if [ -x /etc/rc.local ]; then
	        [ "$VERBOSE" != no ] && log_begin_msg "Running local boot scripts (/etc/rc.local)"
		/etc/rc.local
		ES=$?
		[ "$VERBOSE" != no ] && log_end_msg $ES
		return $ES
	fi
}

case "$1" in
    start)
	do_start
        ;;
    restart|reload|force-reload)
        echo "Error: argument '$1' not supported" >&2
        exit 3
        ;;
    stop|status)
        # No-op
        exit 0
        ;;
    *)
        echo "Usage: $0 start|stop" >&2
        exit 3
        ;;
esac
```

This resulted in about 1% battery drain per hour over.

@mymike00 there is an open issue for UT Tweak Tool (https://gitlab.com/myii/ut-tweak-tool/issues/24) for changing the CPU governor.
